### PR TITLE
Aparna hcl dev

### PR DESCRIPTION
**Pull request by aparnasasidharan2017**
Added 3 containers to HCL

### CODE REVIEW

**Comment by hariharan-devarajan** [comment]
Please consolidate all commits together so that its easier to track.

Is there a way to squash all commits without manually removing diffs? Or should I just fork another branch from this point?
Also, should I merge to SCS/HCL or hariharan-devarajan/HCL?

**Comment by hariharan-devarajan** [comment]
Can you please add a description of what changes you want.

What containers did you add.
Did you add tests.

File: include/hcl/concurrent/skiplist/skiplist_random/Skiplist.h [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+
+  void recycle(NodeType* node) { recycler_.add(node); }
+
+  NodeRecycler<NodeType, NodeAlloc> recycler_;
+  std::atomic<NodeType*> head_;
+  std::atomic<size_t> size_{0};
+};
+
+template <typename T, typename Comp, typename NodeAlloc, int MAX_HEIGHT>
+class ConcurrentSkipList<T, Comp, NodeAlloc, MAX_HEIGHT>::Accessor
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/skiplist/skiplist_random/Skiplist.h]: So as i understand, you have the main data structures defined here and then built HCL container for these. Do the following with these classes, 
- split the two classes in two files which comments on which class does what in detail.
- put files in include/hcl/base/containers/
- keep naming convention similar to other classes. class names and function names are not consistent.

File: include/hcl/concurrent/skiplist/skiplist_random/SkiplistTest.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+   std::vector<struct thread_arg> t_args(num_threads);
+   std::vector<std::thread> workers(num_threads);
+
+   int num_operations = 1000000;
+   int nops = num_operations/num_threads;
+   int rem = num_operations%num_threads;
+
+   auto t1 = std::chrono::high_resolution_clock::now();
+
+   for(int i=0;i<num_threads;i++)
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/skiplist/skiplist_random/SkiplistTest.cpp]: Test need to be moved to test folder. Make sure you test all the APIs within the container. Add this test to the test case of github actions.

File: include/hcl/concurrent/unordered_map/Block_chain_memory.h [hunk 3]
```diff
@@ -0,0 +1,280 @@
+#ifndef __BLOCK_CHAIN_
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/unordered_map/Block_chain_memory.h]: why is this datastructure under unordered map? 

Again if its not extending container move it to include/hcl/base/containers/
fix file name the conversion is not followed correctly,

File: include/hcl/concurrent/unordered_map/Block_chain_memory.h [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+#define __BLOCK_CHAIN_
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/unordered_map/Block_chain_memory.h]: fix header guard convention.

File: include/hcl/concurrent/unordered_map/Block_chain_memory_dist.h [hunk 5]
```diff
@@ -0,0 +1,302 @@
+#ifndef __BLOCK_CHAIN_
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/unordered_map/Block_chain_memory_dist.h]: fix the header guard convention.

File: include/hcl/concurrent/unordered_map/Block_chain_memory.h [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    std::mutex mutex_t;
+    struct node<KeyT,ValueT,HashFcn,EqualFcn> *head;
+};
+
+template <
+    class KeyT,
+    class ValueT, 
+    class HashFcn = std::hash<KeyT>,
+    class EqualFcn = std::equal_to<KeyT>>
+class BlockMap
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/unordered_map/Block_chain_memory.h]: Call this BlockChain but put in include/hcl/base/containers.

File: include/hcl/concurrent/unordered_map/Block_chain_memory_dist.h [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+#define __BLOCK_CHAIN_
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/unordered_map/Block_chain_memory_dist.h]: Fix file name to be just block_chain and put it in its own folder.

File: include/hcl/concurrent/unordered_map/memory_allocation.h [hunk 8]
```diff
@@ -0,0 +1,130 @@
+#ifndef __MEMORY_ALL_H_
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/unordered_map/memory_allocation.h]: What is this memory allocators?

File: include/hcl/concurrent/unordered_map/unordered_map.h [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+#include <float.h>
+#include "Block_chain_memory.h"
+
+namespace hcl {
+
+template <class KeyT, 
+	  class ValueT,
+	  class HashFcn=std::hash<KeyT>,
+	  class EqualFcn=std::equal_to<KeyT>>
+class unordered_map_concurrent : public container
```
Comment (hariharan-devarajan) [inline include/hcl/concurrent/unordered_map/unordered_map.h]: Fix this naming. You should have lock and lock-free folders and they should have unordered_map as a sub-folder

File: test/hashmap_shm_test.cpp [hunk 10]
```diff
@@ -0,0 +1,118 @@
+#include <sys/types.h>
```
Comment (hariharan-devarajan) [inline test/hashmap_shm_test.cpp]: hashmap_shm and hashmap test can they be mergered?

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]
Please check the comments.
- Use this folder structure
  - Within include/hcl your can have lock and lock-free folders
- but all base containers like skip-lists and block-chain in include/hcl/base/container
- then you can have hcl versions in include/hcl/lock or lock-free/skip-lists
- Fix all naming convensions
- Make sure you have all APIs documented and tested in the PR.
- Add the new tests to github actions so that I can check output.
- Squash all commits to one so that its well consolidated.

**Comment by aparnasasidharan2017** [comment]
> Can you please add a description of what changes you want.
> 
> What containers did you add. Did you add tests.

I have added three containers : Unordered Map (hashmap), Queue and Ordered Set(skiplist)
All three are concurrent. The unordered map and skiplist use locks. They are not lockfree. The only lockfree data structure is the concurrent queue from boost with HCL wrapper. But its not fully lockfree, they use locks during dynamic memory allocation. So, basically none of them are lockfree implementations. Do you still want me to create two separate folders or put everything under a single parallel/concurrent folder?

File: include/hcl/concurrent/unordered_map/memory_allocation.h (hunk 8 above)
Comment (aparnasasidharan2017) [reply include/hcl/concurrent/unordered_map/memory_allocation.h]: They are dynamic memory allocators with reuse. They are used by the unordered map and skiplist implementations.

File: test/hashmap_shm_test.cpp (hunk 10 above)
Comment (aparnasasidharan2017) [reply test/hashmap_shm_test.cpp]: Yeah, I'll combine the tests

File: include/hcl/concurrent/unordered_map/Block_chain_memory.h (hunk 3 above)
Comment (aparnasasidharan2017) [reply include/hcl/concurrent/unordered_map/Block_chain_memory.h]: It is a concurrent implementation and not a sequential implementation. Do you want me to change its name?

File: include/hcl/concurrent/skiplist/skiplist_random/SkiplistTest.cpp (hunk 2 above)
Comment (aparnasasidharan2017) [reply include/hcl/concurrent/skiplist/skiplist_random/SkiplistTest.cpp]: Ok

File: include/hcl/concurrent/skiplist/skiplist_random/Skiplist.h (hunk 1 above)
Comment (aparnasasidharan2017) [reply include/hcl/concurrent/skiplist/skiplist_random/Skiplist.h]: Ok

Comment (aparnasasidharan2017) [reply include/hcl/concurrent/skiplist/skiplist_random/Skiplist.h]: Could you check if the naming conventions are okay now?

File: include/hcl/concurrent/skiplist/skiplist_random/SkiplistTest.cpp (hunk 2 above)
Comment (aparnasasidharan2017) [reply include/hcl/concurrent/skiplist/skiplist_random/SkiplistTest.cpp]: I have added the tests to test/CMakeLists.txt. How do I add them to github actions?

**Comment by hariharan-devarajan** [comment]
Why did you close this?

**Comment by hariharan-devarajan** [comment]
I am on vacation till end of Jan that's why I am a little delayed in replying

File: include/hcl/base/containers/concurrent_skiplist/Skiplist-inl.h [hunk 11]
```diff
@@ -0,0 +1,382 @@
+#ifndef HCL_SKIPLIST_INL_H
```
Comment (hariharan-devarajan) [inline include/hcl/base/containers/concurrent_skiplist/Skiplist-inl.h]: Why this name for the file? Can you change - to _ and what is inl?

File: include/hcl/base/containers/concurrent_skiplist/SkiplistAccessor.h [hunk 12]
```diff
@@ -0,0 +1,323 @@
+#ifndef HCL_SKIPLIST_ACCESSOR_H
```
Comment (hariharan-devarajan) [inline include/hcl/base/containers/concurrent_skiplist/SkiplistAccessor.h]: Change file name to skip_list_accessor

File: include/hcl/base/containers/concurrent_skiplist/SkiplistAccessor.h [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+#define HCL_SKIPLIST_ACCESSOR_H
+
+template <typename T, typename Comp, typename NodeAlloc, int MAX_HEIGHT>
+class ConcurrentSkipList<T, Comp, NodeAlloc, MAX_HEIGHT>::Accessor
```
Comment (hariharan-devarajan) [inline include/hcl/base/containers/concurrent_skiplist/SkiplistAccessor.h]: if this is the class Maybe filename should be concurrent_skip_list?

Also for organization have it as concurrent/skip_list/skip_list_inl<full>.h
concurrent/skip_list/skip_list.h
concurrent/skip_list/skip_list_accessor.h

**Review by hariharan-devarajan — COMMENTED** [review]
Please add a test in the workflow yaml.

**Comment by hariharan-devarajan** [comment]
Add a test on ci/run_hcl.sh for your classes so that I can review if the change is working.

**Comment by aparnasasidharan2017** [comment]
I closed it to open another branch with fewer commits. I wasn't sure whether you wanted to merge with SCS/HCL or hariharan-devarajan/HCL. Both branches are the same at the moment.

**Comment by hariharan-devarajan** [comment]
As I understand scs is a fork of my branch, either way once we merge we can
. To do fewer commits u can do a squash of all commits. Google how to
squash commits in git.

**Comment by aparnasasidharan2017** [comment]
When are you back from India? Shall I reopen the PR after your vacation?

**Comment by hariharan-devarajan** [comment]
I will be back on 30th Jan. No you should reopen the PR. I might be slow in
reply but I would be looking at the PR. If I don't reply for more than 3
business days feel free to contact me on email. I will try to update you as
soon as possible.

Best
Hari

**Comment by aparnasasidharan2017** [comment]
Hari,
I have re-opened this PR. I was busy with other things. Hope you had a good vacation. I tried to rebase and squash commits, but it required merging several commits. Btw, just to bring to your notice, two of your testcases failed to compile with intel compiler set - unordered_map_string_test and global_clock_test. Also, the hashmap_test took sometime to run in github actions. It was quite fast on my laptop, when I ran it with > 1 service threads. So, I reduced the default problem size in the testcase. 

Thanks
-Aparna

**Pull request merged**